File: lib/languages/go/loops.go
package main

import (
    "fmt"
    "strings"
)

func main() {
    // Basic for loop with initialization, condition, and post statement
    fmt.Println("Standard for loop:")
    for i := 0; i < 5; i++ {
        fmt.Printf("%d ", i)
    }
    fmt.Println()

    // For loop as a while loop (condition only)
    fmt.Println("\nWhile-style loop:")
    count := 0
    for count < 5 {
        fmt.Printf("%d ", count)
        count++
    }
    fmt.Println()

    // Infinite loop with break
    fmt.Println("\nInfinite loop with break:")
    i := 0
    for {
        if i >= 5 {
            break
        }
        fmt.Printf("%d ", i)
        i++
    }
    fmt.Println()

    // For loop with continue
    fmt.Println("\nLoop with continue (skip even numbers):")
    for i := 0; i < 10; i++ {
        if i%2 == 0 {
            continue
        }
        fmt.Printf("%d ", i)
    }
    fmt.Println()

    // For-range loop over slice
    fmt.Println("\nIterating over a slice:")
    fruits := []string{"apple", "banana", "cherry", "date", "elderberry"}
    for index, fruit := range fruits {
        fmt.Printf("%d: %s\n", index, fruit)
    }

    // For-range with only values (ignoring index)
    fmt.Println("\nIterating over slice values:")
    for _, fruit := range fruits {
        fmt.Println(fruit)
    }

    // For-range with only indices (ignoring values)
    fmt.Println("\nIterating over slice indices:")
    for i := range fruits {
        fmt.Printf("Index %d\n", i)
    }

    // For-range over map
    fmt.Println("\nIterating over a map:")
    ages := map[string]int{
        "Alice": 25,
        "Bob":   30,
        "Carol": 27,
    }
    for name, age := range ages {
        fmt.Printf("%s is %d years old\n", name, age)
    }

    // For-range over string (iterates over runes)
    fmt.Println("\nIterating over a string (runes):")
    for i, char := range "Hello, 世界" {
        fmt.Printf("Character %c at position %d\n", char, i)
    }

    // Nested loops
    fmt.Println("\nNested loops:")
    for i := 1; i <= 3; i++ {
        for j := 1; j <= 3; j++ {
            fmt.Printf("(%d,%d) ", i, j)
        }
        fmt.Println()
    }

    // Labeled break
    fmt.Println("\nLabeled break:")
    outer:
    for i := 0; i < 3; i++ {
        for j := 0; j < 3; j++ {
            if i == 1 && j == 1 {
                fmt.Println("Breaking outer loop")
                break outer
            }
            fmt.Printf("(%d,%d) ", i, j)
        }
        fmt.Println()
    }

    // Labeled continue
    fmt.Println("\nLabeled continue:")
    outer2:
    for i := 0; i < 3; i++ {
        for j := 0; j < 3; j++ {
            if j > i {
                fmt.Println("Continuing outer loop")
                continue outer2
            }
            fmt.Printf("(%d,%d) ", i, j)
        }
        fmt.Println()
    }
}

File: lib/languages/kotlin/loops.kt
// For loop with range
for (i in 1..5) {
       println(i)  // Prints 1, 2, 3, 4, 5
   }

   // For loop with until (exclusive upper bound)
   for (i in 1 until 5) {
       println(i)  // Prints 1, 2, 3, 4
   }

   // For loop with step
   for (i in 1..10 step 2) {
       println(i)  // Prints 1, 3, 5, 7, 9
   }

   // For loop with downTo
   for (i in 5 downTo 1) {
       println(i)  // Prints 5, 4, 3, 2, 1
   }

   // For loop over collection
   val fruits = listOf("apple", "banana", "cherry")
   for (fruit in fruits) {
       println(fruit)
   }

   // For loop with index
   for (i in fruits.indices) {
       println("$i: ${fruits[i]}")
   }

   // For loop with index and value using withIndex
   for (indexedValue in fruits.withIndex()) {
       val i = indexedValue.index
       val fruit = indexedValue.value
       println("$i: $fruit")
   }

   // While loop
   var counter = 0
   while (counter < 5) {
       println(counter)
       counter++
   }

   // Do-while loop
   var number = 10
   do {
       println(number)
       number--
   } while (number > 0)

   // Break statement
   for (i in 1..10) {
       if (i == 5) break
       println(i)  // Prints 1, 2, 3, 4
   }

   // Continue statement
   for (i in 1..10) {
       if (i % 2 == 0) continue
       println(i)  // Prints 1, 3, 5, 7, 9
   }

   // Labeled break
   outer@ for (i in 1..3) {
       for (j in 1..3) {
           if (i == 2 && j == 2) break@outer
           println("$i, $j")
       }
   }

   // Labeled continue
   outer@ for (i in 1..3) {
       for (j in 1..3) {
           if (j == 2) continue@outer
           println("$i, $j")
       }
   }

   // Functional approaches (forEach)
   fruits.forEach { fruit ->
       println(fruit)
   }

   // Functional approaches (map)
   val lengths = fruits.map { it.length }

   // Functional approaches (filter)
   val longFruits = fruits.filter { it.length > 5 }

   // Functional approaches (fold/reduce)
   val totalLength = fruits.fold(0) { acc, fruit -> acc + fruit.length }

   // Sequence for lazy evaluation
   val sequence = (1..1000)
       .asSequence()
       .filter { it % 2 == 0 }
       .map { it * it }
       .take(5)
       .toList()

